7455: Add support for jolokia JMX service connection

skarsaune opened the pull request:
Setting back for review. The azure problem requires a fix in Jolokia.

Make use of support for JMX service connection in jolokia 1.7.0 and later to connect to JVMs over this protocol.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JMC-7455](https://bugs.openjdk.org/browse/JMC-7455): Add support for jolokia JMX service connection (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmc.git pull/332/head:pull/332` \
`$ git checkout pull/332`

Update a local copy of the PR: \
`$ git checkout pull/332` \
`$ git pull https://git.openjdk.org/jmc.git pull/332/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 332`

View PR using the GUI difftool: \
`$ git pr show -t 332`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmc/pull/332.diff">https://git.openjdk.org/jmc/pull/332.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmc/pull/332#issuecomment-979719148)

thegreystone commented on application/org.openjdk.jmc.jolokia/build.properties:
```diff
@@ -0,0 +1,45 @@
+#
+#  Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
```
For new files, don't forget to add the right copyright year (2021), and a line for your current employer (if it makes sense). See Datadog copyrights.

thegreystone commented on application/org.openjdk.jmc.jolokia/plugin.xml:
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--   
+   Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
```
I won't comment regarding copyright in any more files, but the comment before of course applies to all of them.

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/AbstractCachedDescriptorProvider.java:
```diff
@@ -0,0 +1,154 @@
+import org.openjdk.jmc.common.IDescribable;
+import org.openjdk.jmc.rjmx.descriptorprovider.AbstractDescriptorProvider;
+import org.openjdk.jmc.rjmx.descriptorprovider.IDescriptorListener;
+
```
A short class description in public classes is nice. ;)

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/JmcJolokiaJmxConnection.java:
```diff
@@ -0,0 +1,200 @@
+
+/**
+ * 
+ * @author martin@skarsaune.net
```
We typically don't use the author tag in the project.

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/JmcJolokiaPlugin.java:
```diff
@@ -0,0 +1,69 @@
+	
+	public final static String PLUGIN_ID = "org.openjdk.jmc.jolokia"; //$NON-NLS-1$
+
+	// The logger.
```
Comments that don't really add to the understanding, we typically try to avoid.

skarsaune commented on application/org.openjdk.jmc.jolokia/pom.xml:
```diff
@@ -0,0 +1,110 @@
+				</configuration>
+			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
```
Sooo .... I am not sure about the preferred way to pull in external dependencies. I can add that both jolokia libraries come as shaded jars with some OSGi packaging, but I do not know enough about that to say whether that would be a workable or better solution. These are quite new components, so we can probably modify the OSGI packaging in Jolokia going forwards if needed. 
POM: https://github.com/rhuss/jolokia/blob/556b58e63a125fb79d516d7f7833e22270f591b0/client/jmx-adapter/pom.xml#L129
Artifact: https://repo1.maven.org/maven2/org/jolokia/jolokia-jmx-adapter/1.7.0/jolokia-jmx-adapter-1.7.0-standalone.jar

skarsaune replied:
I'm working in a separate branch to expose the dependencies via releng/thirdparty, but have not been able to make it work yet.

skarsaune commented on application/org.openjdk.jmc.kubernetes/pom.xml:
```diff
@@ -0,0 +1,94 @@
+			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-dependency-plugin</artifactId>
```
Same question as for jolokia pom.xml. 
Corresponding resources:
https://github.com/rhuss/jolokia/blob/556b58e63a125fb79d516d7f7833e22270f591b0/client/kubernetes/pom.xml#L44
https://repo1.maven.org/maven2/org/jolokia/jolokia-kubernetes/1.7.0/jolokia-kubernetes-1.7.0-standalone.jar

skarsaune commented on application/tests/org.openjdk.jmc.jolokia.test/src/test/java/org/openjdk/jmc/jolokia/JolokiaTest.java:
```diff
@@ -0,0 +1,163 @@
+	@Test
+	public void testReadAttributesOverJolokia() throws MalformedURLException, IOException, OperationsException,
+			IntrospectionException, AttributeNotFoundException, ReflectionException, MBeanException {
+		for (ObjectName objectName : jolokiaConnection.queryNames(null, null)) {
```
Iterating over all attributes may require some maintenance. However it will also detect any problems with this connector for new JVM developments. 
I can consider just testing a few select attributes instead.

reinhapa commented:
@skarsaune do you have an JMC issue already or do you need help creating one?

skarsaune commented on application/tests/org.openjdk.jmc.jolokia.test/src/test/java/org/openjdk/jmc/jolokia/JolokiaTest.java:
```diff
@@ -0,0 +1,163 @@
+		
+	}
+
+	private static MBeanServerConnection getJolokiaMBeanConnector() throws IOException, MalformedURLException {
```
Testing discovery may pose for CI environments (?). We can consider making it manual if networking is an issue

skarsaune commented:
> 

@reinhapa : Please help creating an issue. I do not have access.

reinhapa commented:
@skarsaune created [JMC-7455](https://bugs.openjdk.java.net/browse/JMC-7455) change the title to: _7455: Add support for jolokia JMX service connection_

skarsaune commented:
I noticed the failed test run. Have used spotbugs to reformat. Re-run please.

skarsaune commented:
I have corrected the last formatting and rebased with master. When I run the shell scripts from the github action definition on my mac, the first 3 run fine, however in scripts/runagenttests.sh there is a test failure (test-permissions). Should not be related to this PR however.

skarsaune commented:
Fails on some platforms: "Unrecognized option: -XstartOnFirstThread" , will fix

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/AbstractCachedDescriptorProvider.java:
```diff
@@ -0,0 +1,158 @@
+import org.openjdk.jmc.rjmx.descriptorprovider.IDescriptorListener;
+
+/**
+ * I cache a list if identified JVMs that can be refreshed in the background by some means of
```
"I" personal pronomen. "if" identified. 

Better to keep it impersonal and describe what the class does. E.g. "Keeps a list of identified JVMs..."
"The {@code AbstractCachedDescriptorProvider} keeps a list of identified JVMs..." would be even better.

skarsaune replied:
I see. Will adjust 😄 . FYI: I'm on the old Smalltalk extreme object oriented lingo. You may have seen it in SWT and JFace code as well. A bit out of date. 
I'm currently on 2 week holiday with family, will get back to the PR next year 🌴

thegreystone replied:
Enjoy your vacation! :)

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/JmcJolokiaJmxConnection.java:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	/**
+	 * build / reverse engineer MBeanOperationInfo by using the local one if it is a match or try to
```
Punctuation.

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/JmcJolokiaJmxConnection.java:
```diff
@@ -0,0 +1,198 @@
+	 * @return Descriptor
+	 */
+	private MBeanOperationInfo stealOrBuildOperationInfo(MBeanOperationInfo original, MBeanInfo localInfo) {
+		// first attempt to get descriptor from local copy
```
Might want to break out into individual methods at the comment lines.

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/JolokiaAgentDescriptor.java:
```diff
@@ -0,0 +1,200 @@
+import org.openjdk.jmc.ui.common.jvm.JVMDescriptor;
+import org.openjdk.jmc.ui.common.jvm.JVMType;
+
+public class JolokiaAgentDescriptor implements ServerConnectionDescriptor {
```
Public class. Would prefer a class comment.

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/JolokiaAgentDescriptor.java:
```diff
@@ -0,0 +1,200 @@
+
+	/**
+	 * Best effort to extract JVM information from a connection if everything works. Can be adjusted
+	 * to support different flavors of JVM
```
Punctuation, last sentence.

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/JolokiaDiscoveryListener.java:
```diff
@@ -0,0 +1,94 @@
+import org.openjdk.jmc.ui.common.jvm.JVMDescriptor;
+import org.openjdk.jmc.jolokia.preferences.PreferenceConstants;
+
+public class JolokiaDiscoveryListener extends AbstractCachedDescriptorProvider implements PreferenceConstants {
```
Class comment.

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/ServerConnectionDescriptor.java:
```diff
@@ -0,0 +1,48 @@
+import org.openjdk.jmc.rjmx.IServerDescriptor;
+
+/**
+ * Describes the JVM and how to connect to it
```
Punctuation.

thegreystone commented on application/org.openjdk.jmc.jolokia/src/main/java/org/openjdk/jmc/jolokia/preferences/PreferenceConstants.java:
```diff
@@ -0,0 +1,43 @@
+package org.openjdk.jmc.jolokia.preferences;
+
+/**
+ * Constant definitions for plug-in preferences
```
Punctuation.

thegreystone commented on application/org.openjdk.jmc.kubernetes/src/main/java/org/openjdk/jmc/kubernetes/JmcKubernetesJmxConnection.java:
```diff
@@ -0,0 +1,80 @@
+import org.openjdk.jmc.jolokia.JmcJolokiaJmxConnection;
+import org.openjdk.jmc.rjmx.ConnectionException;
+
+public class JmcKubernetesJmxConnection extends JmcJolokiaJmxConnection {
```
Class comment.

thegreystone commented on application/org.openjdk.jmc.kubernetes/src/main/java/org/openjdk/jmc/kubernetes/JmcKubernetesJmxConnectionProvider.java:
```diff
@@ -0,0 +1,52 @@
+import javax.management.remote.JMXConnectorProvider;
+import javax.management.remote.JMXServiceURL;
+
+public class JmcKubernetesJmxConnectionProvider implements JMXConnectorProvider {
```
Class comment. E.g. "This {@code JMXConnectionProvider} verifies that the protocol is kubernetes."

thegreystone commented on application/org.openjdk.jmc.kubernetes/src/main/java/org/openjdk/jmc/kubernetes/KubernetesDiscoveryListener.java:
```diff
@@ -0,0 +1,323 @@
+import io.fabric8.kubernetes.client.internal.KubeConfigUtils;
+import io.fabric8.kubernetes.client.utils.Utils;
+
+public class KubernetesDiscoveryListener extends AbstractCachedDescriptorProvider {
```
Class comment.

thegreystone commented on application/org.openjdk.jmc.kubernetes/src/main/java/org/openjdk/jmc/kubernetes/preferences/JmcKubernetesPreferenceForm.java:
```diff
@@ -0,0 +1,133 @@
+ * belongs to the main plug-in class. That way, preferences can be accessed directly via the
+ * preference store.
+ */
+
```
Line 57 is superfluous.

thegreystone commented on application/org.openjdk.jmc.kubernetes/src/main/java/org/openjdk/jmc/kubernetes/preferences/PreferenceConstants.java:
```diff
@@ -0,0 +1,50 @@
+package org.openjdk.jmc.kubernetes.preferences;
+
+/**
+ * Constant definitions for plug-in preferences
```
Punctuation.

thegreystone reviewed (changes requested):
I took a quick pass on (mainly) the javadocs.

skarsaune commented:
It would be good to get a new CI run with the new way of handling OSGi dependencies

skarsaune commented on application/tests/org.openjdk.jmc.jolokia.test/pom.xml:
```diff
@@ -0,0 +1,116 @@
+			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-dependency-plugin</artifactId>
```
Wondering whether I should do the same OSGi packaging solution with third party and platform definitions here. But is it ok to include test dependencies there as well, or should they be kept out of it?

skarsaune replied:
https://jdkmissioncontrol.slack.com/archives/CGMRA4ALW/p1643685017130109

skarsaune commented:
Hi again! Have been missing in action for a bit due to very critical phase at client, but have taken some time off my holiday to figure out a way to test the kubernetes support by mocking. If I could please have a CI test run to see that it works fine in the pipeline. Also rebased and updated to 8.3. cc: @thegreystone

skarsaune commented:
Yeah, whenever I think I am done done, I think of something more to fix .... Set to draft now. Will probably draw the line in not too many days.

thegreystone commented:
Hi @skarsaune! The master branch is now JMC 9, and now would be a great opportunity to get this in. Is there anything else you'd like to fix before I look at it again?

skarsaune commented:
Great @thegreystone . Some actions I must take:
1. Align with current master
2. Do some manual testing in jmc after updating
3. Ideally I would like to see if I can get a new release of Jolokia with some improvements

skarsaune commented:
Manual tests look good. Solved a problem with stored credentials. Have requested merges and a release on the jolokia side.
Still problems running the test though. Would it be possible to request a CI run on this PR?

thegreystone commented:
Hi @skarsaune! Are you ready to make this PR ready for review? 

Regarding the tests, typically the tests would be run in your fork, on commits to your branch. I can try running them on a few OSes locally here though.

skarsaune commented:
1. Great if you can run the tests @thegreystone ! The jolokia tests run , but the Kubernetes tests not. They should be set up the same way, but it here is probably a silly error somewhere.

2. Thanks for checking in ☺️. I just asked for a release of jolokia. I hope to conclude that soon.

skarsaune commented:
Jolokia is updated now. Merged with master and marking ready for review. PR is ready for slaughter 😉

skarsaune commented:
Found a functional problem with the Kubernetes test after upgrading to jolokia 1.7.2. will investigate ASAP

skarsaune commented:
@thegreystone : I believe the previous review should be addressed. You can have a new look when you have time, and any assistance on running the test would be appreciated.

skarsaune commented:
Still troubleshooting Kubernetes tests. Now they run , but fail with class not found on Apache commons http classes

thegreystone commented:
Hi Martin! We're getting closer to the JMC release date. Is this something you're aiming to get into JMC 9?

RoiSoleil commented:
Really interested by the jolokia support in JMC. Will be amazing to have it in the next release ;)

skarsaune commented:
Sorry, missed this email. I'm still struggling with the classloading of the
kubernetes tests, and it has been hard to prioritze lately. It should be
reproducible running `./build.sh --test` on both linux and macos.

On Wed, Jan 3, 2024 at 7:26 PM Marcus Hirt ***@***.***> wrote:

> Hi Martin! We're getting closer to the JMC release date. Is this something
> you're aiming to get into JMC 9?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openjdk/jmc/pull/332#issuecomment-1875787578>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQB2GSKG654E3LAYEABNSLYMWPDHAVCNFSM5GYUGXXKU5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TCOBXGU3TQNZVG44A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

skarsaune commented:
replaced by https://github.com/openjdk/jmc/pull/548
